Allow to build with cabal

hasufell opened the pull request:
The process to create these files is roughly:

1. run `stack2cabal`
2. work around cabal bugs (see https://github.com/haskell/cabal/issues/5444)
3. disable tests to avoid build failures with --enable-tests from
   unrelated packages (cabal considers all packages defined via
   source-repository-package as *local* and will try to build its tests)
4. fix package constraints in cabal.project.freeze, where stackage
   constraints were overwritten in stack.yaml files (usually raising
   a version number)

Rationale:

1. good to support all available tools as options for interested users
2. cabal allows easier managing of package bounds (and testing against
   latest package versions)
3. #1303 is an excellent workaround for nix users, but requires nix.
   This should be the go-to fallback in case cabal project files
   get out of sync for too long.

Commit:
Allow to build with cabal

The process to create these files is as follows:

1. Run stack2cabal
2. Disable tests and benchmarks of local packages
   (those specified via source-repository-package...
   this won't be needed with cabal-3.4 anymore, because it won't treat
   them as local anymore)
3. Hackage packages in `packages` in `cardano-1.19.x.yaml` or `extra-deps`
   in `stack.yaml`, which are specified as git repositories might
   have different versions than the stack resolver. `stack2cabal`
   cannot pick these up, so these must be adjusted manually (or deleted) in
   `cabal.project.freeze` (at the time of writing: http-client and persistent).

Rationale:

1. good to support all available tools as options for interested users
2. cabal allows easier managing of package bounds (and testing against
   latest package versions and different GHCs, because it is not
   constrained to stackage)
3. doesn't require nix, which is an additional step to set up and
   maintain and requires significant amount of disk space

hasufell commented on cabal.project:
```diff
@@ -1,2 +1,438 @@
+
+-- force systemd off until there's a solution to
+-- https://github.com/haskell/cabal/issues/5444
+flags: -systemd
```
This is being debated in https://github.com/input-output-hk/cardano-node/pull/1666 as well

Commit:
Fix build for cabal users that don't have systemd

This is mainly due to a cabal bug and should be removed in
the future: https://github.com/haskell/cabal/issues/5444

There are various distros, where installing development files
of libsystemd won't be an option:

- Alpine (doesn't have systemd at all)
- Gentoo (not installable when eudev is installed)
- Exherbo (not installable when eudev is installed)

Users are expected to run the script and then execute cabal like so:

  cabal build --project-file=cabal.nosystemd.project all

KtorZ commented on gen-cabal-nosystemd.sh:
```diff
@@ -0,0 +1,14 @@
+# force systemd off until there's a solution to
+# https://github.com/haskell/cabal/issues/5444
+
+set -eux
```
let's move this one under "scripts" :+1:

Commit:
Fix build on windows

Also see https://github.com/phadej/time-compat/issues/10

Commit:
Pin GHC to 8.6.5 to avoid confusion

KtorZ committed:
add buildkite check to make sure that the cabal.project stays up-to-date

rvl committed:
nix: Add a shell file which can be used for plain cabal builds

The Haskell.nix shellFor function provides Haskell package
dependencies and all their dependencies.

This new nix/cabal-shell.nix file only provides a compiler, build
tools and system libraries, and lets Cabal build the Haskell package
dependencies.

rvl committed:
Add an alternate Cabal project file for nix+cabal builds

Using nix-shell and "cabal build --project-file=cabal-nix.project",
all Haskell dependencies will be downloaded from the IOHK cache,
rather than built locally.

rvl committed:
Hide cabal.project.freeze in GitHub PR diffs
